lib/segment/src/index/hnsw_index/hnsw.rs

```diff
@@ -668,16 +717,22 @@ impl<TGraphLinks: GraphLinks> VectorIndex for HNSWIndex<TGraphLinks> {
         let payload_m = self.config.payload_m.unwrap_or(self.config.m);
 
         if payload_m > 0 {
+            // Calculate true average number of links per vertex in the HNSW graph
+            // to better estimate percolation threshold
+            let average_links_per_0_level =
+                graph_layers_builder.get_average_connectivity_on_level(0);
+            let average_links_per_0_level_int = (average_links_per_0_level as usize).max(1);
+
             for (field, _) in payload_index.indexed_fields() {
                 debug!("building additional index for field {}", &field);
 
                 // It is expected, that graph will become disconnected less than
                 // $1/m$ points left.
                 // So blocks larger than $1/m$ are not needed.
                 // We add multiplier for the extra safety.
-                let percolation_multiplier = 2;
+                let percolation_multiplier = 4;
                 let max_block_size = if self.config.m > 0 {
-                    total_vector_count / self.config.m * percolation_multiplier
+                    total_vector_count / average_links_per_0_level_int * percolation_multiplier
```

Comment:
This pair of changes is not directly related to the PR, but it helps to test changes more easily, so I decided to include them anyway.
The original problem is, that apparently current thresholds are too weak to address precision loss in out test dataset, so this changes here should improve performance in similar cases. To better understand which values are the most optimal, some additional experiments should be conducted.


The first change, `percolation_multiplier` should allow larger payload blocks to create an additional links in HNSW graph. Original multiplier value was selected randomly, so changing it to another arbitrary values, even with just anecdotal evidence of effectiveness, should be OK.

Second change  - `average_links_per_0_level_int` improves threshold prediction by measuring actual number of average connection per node, which might be less than expected due to HNSW heuristic that we use.